geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java

```diff
@@ -1886,12 +1826,49 @@ private long getUniqueId(Principal principal) throws Exception {
           securityLogWriter.warning(
               String.format(
                   "%s: Post-process authorization enabled, but no authentication callback (%s) is configured",
-                  new Object[] {getName(), SECURITY_CLIENT_AUTHENTICATOR}));
+                  getName(), SECURITY_CLIENT_AUTHENTICATOR));
         }
       }
       postAuthzRequest =
           new AuthorizeRequestPP(postAuthzFactoryName, getProxyID(), principal, getCache());
     }
     return setUserAuthorizeAndPostAuthorizeRequest(authzRequest, postAuthzRequest);
   }
+
+  static class ProcessingMessageTimer {
+    public static final long NOT_PROCESSING = -1L;
+
+    private final AtomicLong processingMessageStartTime = new AtomicLong(NOT_PROCESSING);
+
+    /**
+     * Set or resets time regardless if already set.
+     */
+    void setProcessingMessage() {
+      processingMessageStartTime.set(System.currentTimeMillis());
+    }
+
+    /**
+     * Updates time if previously set.
+     */
+    void updateProcessingMessage() {
+      long current;
+      while ((current = processingMessageStartTime.get()) != NOT_PROCESSING) {
```

Comment:
If some other thread changes the atomic between lines 1855 and 1856 then they changed it to either -1 or System.currentTimeMillis. If either of these changes happen then we don't need to retry. If they changed it to -1 then this method is done. If they changed it to currentTimeMillis then they did for us what we are trying to do. So I think this code can be safely changed to:
long current = processingMessageStartTime.get();
if (current != NOT_PROCESSING) {
  processingMessageStartTime.compareAndSet(current, System.currentTimeMillis());
}
Under high contention this will improve performance since we can have multiple threads constantly changing it.
One more optimization. After the above "if" you could add:
  long now = System.currentTimeMillis();
  if (now != current) {
     processingMessageStartTime.compareAndSet(current, now);
  }
Since "currentTimeMillis" may not change every millisecond this could reduce the number of compareAndSet calls